packages/firestore/src/local/tab_notification_channel.ts

```diff
@@ -0,0 +1,393 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Code, FirestoreError } from '../util/error';
+import { BatchId, InstanceKey, TargetId } from '../core/types';
+import { assert } from '../util/assert';
+import { debug, error } from '../util/log';
+import { primitiveComparator } from '../util/misc';
+import { SortedSet } from '../util/sorted_set';
+import {
+  validateNamedArray,
+  validateNamedOptionalType,
+  validateNamedType
+} from '../util/input_validation';
+
+// Prefix keys used in WebStorage.
+const FIRESTORE_PREFIX = 'fs';
+const INSTANCE_ROW_PREFIX = 'instances';
+
+const LOG_TAG = 'TabNotificationChannel';
+
+/**
+ * WebStorage of the Firestore client.
+ *
+ * Firestore uses WebStorage for cross-tab notifications to propagate the
+ * state of pending mutation batches and query targets. All Firestore tabs are
+ * responsible for persisting their instance state as their in-memory state
+ * changes. Primary tabs will further update the state of each mutation batch
+ * and query target.
+ */
+export interface TabNotificationChannel {
+  /** Associates a new Mutation Batch with the current Firestore instance. */
+  addPendingMutation(batchId: BatchId): void;
+
+  /** Removes a Mutation Batch from the current Firestore instance. */
+  removePendingMutation(batchId: BatchId): void;
+
+  /** Gets the minimum mutation batch for all active instances. */
+  getMinPendingMutation(): BatchId | null;
+
+  /** Associates a new Query Target with the current Firestore instance. */
+  addActiveQueryTarget(targetId: TargetId): void;
+
+  /** Removes a new Query Target from the current Firestore instance. */
+  removeActiveQueryTarget(targetId: TargetId): void;
+
+  /** Gets the active query targets for all active instances. */
```

Comment:
I thought about this and don't think it's going to have a noticeable performance impact. This will only get called when a new primary tab gets selected.

I added a comment here and to `getMinimumGlobalPendingMutation` about their runtime complexity.